luigi/contrib/jupyter_notebook.py

```diff
@@ -0,0 +1,310 @@
+
+**The jupyter_notebook module depends on both the nbconvert (>=5.3.1)
+and the nbformat (>=4.4.0) modules. Please make sure they are installed.**
+
+Written by `@mattiaciollaro <https://github.com/mattiaciollaro>`_.
+"""
+
+import json
+import logging
+import os
+
+import luigi
+
+logger = logging.getLogger('luigi-interface')
+
+try:
+    import nbformat
+    from nbconvert.preprocessors import (
+        CellExecutionError,
+        ExecutePreprocessor
+    )
+
+except ImportError:
+    logger.warning('Loading jupyter_notebook module without nbconvert '
+                   'and/or nbformat installed. The nbconvert and nbformat '
+                   'modules are required to use the jupyter_notebook module. '
+                   'Please install nbconvert and nbformat.')
+
+
+def _get_file_name_from_path(input_path):
+    """
+    A simple utility to extract the name of a file without the file extension
+    from a given path.
+
+    This function extracts the file name without the file extension from a given
+    path.
+    For example, if `path=~/some_dir/some_file.ext`, then
+    `get_filename_from_path(path)` will return `some_file`.
+
+    :param input_path: a path to a file
+
+    :returns: a string containing the name of the file without the file
+        extension
+    """
+    base_name = os.path.basename(os.path.normpath(input_path))
+    file_name = base_name.split('.')[0]
+    return file_name
+
+
+def _get_values(obj):
+    """
+    A simple utility to extract the values of a dictionary, list, or other
+    iterable and recombine them into a list.
+
+    :param obj: a dictionary, list, or other iterable
+
+    :returns: a list with the values of obj
+    """
+    if isinstance(obj, dict):
+        out = obj.values()
+    else:
+        out = [val for val in obj]
+    return out
+
+
+class JupyterNotebookTask(luigi.Task):
+    """
+    This is a template task to execute Jupyter notebooks as Luigi tasks.
+    This task has the following key parameters:
+
+    :param nb_path: the full path to the Jupyter notebook (**required**).
+
+    :param kernel_name: the name of the kernel to be used in the notebook
+        execution (**required**).
+
+    :param timeout: maximum time (in seconds) allocated to run each cell.
+        If -1 (the default), no timeout limit is imposed.
+
+    :param pars: a dictionary of user-defined parameters to be passed to the
+        Jupyter notebook.
+
+    :param json_action: if `delete` (default), the temporary JSON file with
+        **pars** is deleted at the end of the task execution; if `keep`, the
+        temporary JSON file is kept (useful for debugging purposes).
+    """
+    nb_path = luigi.Parameter(
+        default=None
+    )
+
+    kernel_name = luigi.Parameter(
+        default=None
+    )
+
+    timeout = luigi.IntParameter(
+        default=-1
+    )
+
+    json_action = luigi.Parameter(
+        default='delete'
+    )
+
+    pars = {}
+
+    def _form_input(self):
+
```

Comment:
The `_form_input` and `_form_output` methods are a bit hard to read due to the lambdas, so I would consider adding a short docstring to each.

Reply:
I documented `_form_input` and `_form_output` with a doctoring that explains the logic behind them.